### PR TITLE
Scope Hoisting: Don't insert unused requires that aren't registered anywhere

### PR DESCRIPTION
**Pull request by gorakong**
# ↪️ Pull Request

In `getSymbolResolution`,  unused modules can end up in an asset's hoisted requires. A parcelRequire is then inserted for the unused module, and results in a `Cannot find module` runtime error (see example below).

Proposed fix is to check that an asset's exportSymbols are actually used before inserting a parcelRequire for it.

Tested on a large app.

## 💻 Examples

```
$parcel$export(module.exports, "foo", () => (parcelRequire("jDXqh")).foo);

var $jDXqh = parcelRequire("jDXqh");
var $5oNdt = parcelRequire("5oNdt");                      // 5oNdt isn't registered anywhere
eval('');

});
parcelRequire.register("jDXqh", function(module, exports) {

$parcel$export(module.exports, "foo", () => $e4cfc0f7eb90214f$export$6a5cdcad01c973fa);
const $e4cfc0f7eb90214f$export$6a5cdcad01c973fa = 'foo';

});
```

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.32s | +17.00ms |
| Cached | 464.00ms | -17.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.58s | -115.00ms |
| Cached | 274.00ms | +12.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/329511124623950420)**

File: packages/packagers/js/src/ScopeHoistingPackager.js [hunk 1]
```diff
@@ -741,10 +741,22 @@ ${code}
...
+      let resolvedAssetUsedSymbols = nullthrows(
+        this.bundleGraph.getUsedSymbols(resolvedAsset),
       );
+
+      if (
+        resolvedAssetUsedSymbols.has(exportSymbol) ||
+        nullthrows(this.bundleGraph.getUsedSymbols(dep)).has(exportSymbol) ||
+        (exportSymbol === 'default' &&
+          resolvedAssetUsedSymbols.has(exportSymbol)) ||
+        (exportSymbol === '*' && resolvedAssetUsedSymbols.size > 0)
```
Comment (wbinnssmith) [inline packages/packagers/js/src/ScopeHoistingPackager.js]: Maybe a comment (or temporary variable) that explains this large expression will help 🙂

**Comment by wbinnssmith** [comment]
The test looks great, but I'll defer to @devongovett or @mischnic to have a look at the scope hoisting logic.

File: packages/packagers/js/src/ScopeHoistingPackager.js [hunk 2]
```diff
@@ -741,10 +741,28 @@ ${code}
...
+      // Check to see if the exportSymbol is in the resolved asset or the dependency's
+      // set of used symbols. If the exportSymbol is 'default', check that the
+      // default export from the resolved asset is used.
+      // If the exportSymbol is '*', a non-empty set of the resolved asset's used symbols
+      // indicates the export(s) is used.
+      let isExportSymbolUsed =
+        resolvedAssetUsedSymbols.has(exportSymbol) ||
+        nullthrows(this.bundleGraph.getUsedSymbols(dep)).has(exportSymbol) ||
+        (exportSymbol === 'default' &&
+          resolvedAssetUsedSymbols.has(exportSymbol)) ||
```
Comment (devongovett) [inline packages/packagers/js/src/ScopeHoistingPackager.js]: Doesn't the first condition already cover this?

Comment (gorakong) [reply packages/packagers/js/src/ScopeHoistingPackager.js]: the second condition covers the case where a default export is re-exported as some named export:
```
// a.js
import {foo} from './b.js';

// b.js
export {default as foo} from './c.js';

// c.js
export default function(name) { ... }
```
`foo` is in the usedSymbols set of the './b' dep on a.js (whereas c's usedSymbols only contains `default`)

Comment (devongovett) [reply packages/packagers/js/src/ScopeHoistingPackager.js]: You have `resolvedAssetUsedSymbols.has(exportSymbol)` on line 754, which implies that `exportSymbol === 'default' && resolvedAssetUsedSymbols.has(exportSymbol)` is true, right?

Comment (gorakong) [reply packages/packagers/js/src/ScopeHoistingPackager.js]: true! that condition's redundant and has been removed.

**Comment by wbinnssmith** [comment]
What's the state of this PR?

**Comment by mischnic** [comment]
~So just checking `resolvedAssetUsedSymbols.has(exportSymbol)` would also fix the integration test. (And that is also what I would have assumed to work). Can you add tests (or extend the existing one by adding more `library` folders) so that these conditions are also convered? (Related to https://github.com/parcel-bundler/parcel/pull/7764#discussion_r827477360)~ Ah, that breaks existing tests. I see...

**Comment by mischnic** [comment]
This very long condition made me suspicious and indeed the following version passes all test (you probably want to test this in your app though).
The new condition also better corresponds to the cases where it shouldn't insert the statement, which is when no `parcelRequire.register` was generated (= skipped in this bundle and also not imported from another bundle)

```js
  if (
    isWrapped &&
    dep &&
    !dep?.meta.shouldWrap &&
    symbol !== false &&
    // Only do this if the asset is part of a different bundle (so it was definitely
    // parcelRequire.register'ed there), or if it is indeed registered in this bundle.
    (!this.bundle.hasAsset(resolvedAsset) ||
      !this.shouldSkipAsset(resolvedAsset))
  ) {
    let hoisted = this.hoistedRequires.get(dep.id);
    if (!hoisted) {
      hoisted = new Map();
      this.hoistedRequires.set(dep.id, hoisted);
    }

    hoisted.set(
      resolvedAsset.id,
      `var $${publicId} = parcelRequire(${JSON.stringify(publicId)});`,
    );
  }
```

**Comment by gorakong** [comment]
@devongovett updated with @mischnic's suggestion and tested on a large app -- everything looks good 👍

**Review by devongovett — APPROVED** [review]

**Pull request merged**